web/packages/url-utils/src/index.ts

```
@@ -0,0 +1,91 @@
+	RESTRICTED = 'r',
+	X = 'x',
+}
+
+interface GravatarAvatarOptions {
+	default?: GravatarDefault | string;
+	size?: number;
+	rating?: GravatarRating;
+	forceDefault?: boolean;
+}
+
+// Function to get the SHA256 hash of the email
+function getHash( email: string ): string {
+	if ( typeof email !== 'string' || email.trim() === '' ) {
+		throw new Error( 'Valid email is required' );
+	}
+	return sha256( email.trim().toLowerCase() );
+}
+
+// Function to generate the query string from options
+function getQueryString( options: GravatarAvatarOptions ): string {
+	const params: { [ key: string ]: string } = {};
+	for ( const key in options ) {
+		if ( key === 'size' ) {
+			params.size = options.size.toString();
+		} else if ( key === 'default' ) {
+			const defaultValue = options.default;
+			params.default = defaultValue;
+		} else if ( key === 'rating' ) {
+			const ratingValue = options.rating;
+			params.rating = ratingValue;
+		} else if ( key === 'forceDefault' ) {
+			params.forcedefault = options.forceDefault ? 'y' : 'n';
+		}
+	}
+
+	const queryParams = Object.keys( params )
+		.map( ( key ) => {
+			const value = params[ key ];
+			return encodeURIComponent( key ) + '=' + encodeURIComponent( String( value ) );
+		} )
+		.join( '&' );
+	return queryParams ? `?${ queryParams }` : '';
+}
+
+// Generate the Gravatar image URL using SHA256 hash
+// Documentation: https://docs.gravatar.com/api/avatars/images/
+export function avatarUrl( email: string, options: GravatarAvatarOptions = {} ): string {
+	const hash = getHash( email );
+	const query = getQueryString( options );
+	return `${ AVATAR_BASE_URL }${ hash }${ query }`;
+}
+
+// Generate the Gravatar profile URL using SHA256 hash
+// Documentation: https://docs.gravatar.com/api/profiles/
+export function profileUrl( email: string, format: GravatarFormat = GravatarFormat.HTML ): string {
+	const hash = getHash( email );
+	const formatString = format === GravatarFormat.HTML ? '' : `.${ format }`;
```

Q: I'm wondering whether we need to validate the `format`? Are we intentionally exposing the error to developers? (I'm OK with it)

---

Good point, what's the usual behavior for this kind of typescript library that could be used in a typescript or javascript project?

If we don't validate the format, JS users might get a silent error, is that right?

---

It's more about the result of the generated URL. Parameters with invalid values will eventually lead to an error on Gravatar.com, where the user will see a 404 page. They can then correct the invalid values themselves, which is fine.

---

Since this library doesn't do much, I added parameter validation in 1cb58258afb7b240dc1d82be1943988f327f30e0